src/hotspot/share/classfile/resolutionErrors.cpp

```diff
@@ -25,162 +25,131 @@
-  if (_message != NULL) {
+  if (_message != nullptr) {
     _message->increment_refcount();
   }
-}
 
-void ResolutionErrorEntry::set_cause(Symbol* c) {
-  _cause = c;
-  if (_cause != NULL) {
+  if (_cause != nullptr) {
     _cause->increment_refcount();
   }
-}
 
-void ResolutionErrorEntry::set_cause_msg(Symbol* c) {
-  _cause_msg = c;
-  if (_cause_msg != NULL) {
+  if (_cause_msg != nullptr) {
     _cause_msg->increment_refcount();
   }
 }
 
-void ResolutionErrorEntry::set_nest_host_error(const char* message) {
-  _nest_host_error = message;
-}
-
-void ResolutionErrorTable::free_entry(ResolutionErrorEntry *entry) {
+ResolutionErrorEntry::~ResolutionErrorEntry() {
   // decrement error refcount
-  if (entry->error() != NULL) {
-    entry->error()->decrement_refcount();
+  if (error() != NULL) {
+    error()->decrement_refcount();
   }
-  if (entry->message() != NULL) {
-    entry->message()->decrement_refcount();
+  if (message() != NULL) {
+    message()->decrement_refcount();
   }
-  if (entry->cause() != NULL) {
-    entry->cause()->decrement_refcount();
+  if (cause() != NULL) {
+    cause()->decrement_refcount();
   }
-  if (entry->cause_msg() != NULL) {
-    entry->cause_msg()->decrement_refcount();
+  if (cause_msg() != NULL) {
+    cause_msg()->decrement_refcount();
   }
-  if (entry->nest_host_error() != NULL) {
-    FREE_C_HEAP_ARRAY(char, entry->nest_host_error());
+  if (nest_host_error() != NULL) {
+    FREE_C_HEAP_ARRAY(char, nest_host_error());
   }
-  BasicHashtable<mtClass>::free_entry(entry);
 }
 
+class ResolutionErrorDeleteIterate : StackObj {
+private:
+  ConstantPool* p;
 
-// create resolution error table
-ResolutionErrorTable::ResolutionErrorTable(int table_size)
-    : Hashtable<ConstantPool*, mtClass>(table_size, sizeof(ResolutionErrorEntry)) {
-}
+public:
+  ResolutionErrorDeleteIterate(ConstantPool* pool):
+    p(pool) {};
 
-// RedefineClasses support - remove matching entry of a
-// constant pool that is going away
+  bool do_entry(uintptr_t key, ResolutionErrorEntry* value){
+    return (value -> pool() == p);
+  }
+};
+
+// Delete entries in the table that match with ConstantPool c
 void ResolutionErrorTable::delete_entry(ConstantPool* c) {
   assert_locked_or_safepoint(SystemDictionary_lock);
-  for (int i = 0; i < table_size(); i++) {
-    for (ResolutionErrorEntry** p = bucket_addr(i); *p != NULL; ) {
-      ResolutionErrorEntry* entry = *p;
-      assert(entry->pool() != NULL, "resolution error table is corrupt");
-      if (entry->pool() == c) {
-        *p = entry->next();
-        free_entry(entry);
-      } else {
-        p = entry->next_addr();
-      }
-    }
-  }
+
+  ResolutionErrorDeleteIterate deleteIterator(c);
+  _resolution_error_table.unlink(&deleteIterator);
 }
 
+class ResolutionIteratePurgeErrors : StackObj {
+public:
+  bool do_entry(uintptr_t key, ResolutionErrorEntry* value) {
+    ConstantPool* pool = value -> pool();
+    return !(pool->pool_holder()->is_loader_alive());
```

Comment:
The `ResolutionErrorEntry` also needs to be freed. I think it can be done like this:

```
if (!(pool->pool_holder()->is_loader_alive())) {
  delete value;
  return true;
} else {
  return false;
}
```

You can add some `tty->print_cr` in `ResolutionErrorEntry::~ResolutionErrorEntry()` to verify that the destructor is actually called (I believe it's not called with the current version of this PR). Or, you can set a breakpoint there inside a debugger.

Similar changes are needed in `ResolutionErrorDeleteIterate`